File: backend/app/routers/leaderboard.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.database import get_db
from app.models.user import User
from app.schemas.game import MemoryLeaderboardEntry
from app.schemas.match import ArenaLeaderboardEntry
from app.schemas.ranking import RankingMetric, RankingOut
from app.services import match_service, memory_game_service, ranking_service

router = APIRouter(prefix="/leaderboard", tags=["leaderboard"])


@router.get("/arena", response_model=list[ArenaLeaderboardEntry])
async def leaderboard_arena(db: AsyncSession = Depends(get_db), _user: User = Depends(get_current_user)):
    return await match_service.arena_leaderboard(db)


@router.get("/memory", response_model=list[MemoryLeaderboardEntry])
async def leaderboard_memory(db: AsyncSession = Depends(get_db), _user: User = Depends(get_current_user)):
    return await memory_game_service.leaderboard(db)


@router.get("/ranking", response_model=RankingOut)
async def leaderboard_ranking(
    metric: RankingMetric, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    return await ranking_service.get_ranking(db, metric, user)

File: frontend/src/api/leaderboard.ts
import { api } from "@/lib/api";
import type { ArenaLeaderboardEntry, MemoryLeaderboardEntry, RankingMetric, RankingResult } from "@/types";

export async function fetchArenaLeaderboard(): Promise<ArenaLeaderboardEntry[]> {
  const { data } = await api.get<ArenaLeaderboardEntry[]>("/leaderboard/arena");
  return data;
}

export async function fetchMemoryLeaderboard(): Promise<MemoryLeaderboardEntry[]> {
  const { data } = await api.get<MemoryLeaderboardEntry[]>("/leaderboard/memory");
  return data;
}

export async function fetchRanking(metric: RankingMetric): Promise<RankingResult> {
  const { data } = await api.get<RankingResult>("/leaderboard/ranking", { params: { metric } });
  return data;
}
